signac/contrib/migration/__init__.py

```diff
@@ -5,102 +5,135 @@
         )
 
-    while get_config_schema_version() < schema_version:
-        for (origin, destination), migration in MIGRATIONS.items():
-            if version.parse(origin) == get_config_schema_version():
+    guess = current_schema_version
+    while _get_config_schema_version(root_directory, guess) < schema_version:
+        for (origin, destination), migration in _MIGRATIONS.items():
+            if version.parse(origin) == _get_config_schema_version(
+                root_directory, guess
+            ):
                 yield (origin, destination), migration
+                guess = destination
                 break
         else:
             raise RuntimeError(
                 "The signac schema version used by this project is {}, but signac {} "
                 "uses schema version {} and does not know how to migrate.".format(
-                    get_config_schema_version(), __version__, schema_version
+                    _get_config_schema_version(root_directory, guess),
+                    __version__,
+                    schema_version,
                 )
             )
 
 
-def apply_migrations(project):
+def apply_migrations(root_directory):
     """Apply migrations to a project."""
-    with _lock_for_migration(project):
-        for (origin, destination), migrate in _collect_migrations(project):
-            try:
-                print(
-                    f"Applying migration for version {origin} to {destination}... ",
-                    end="",
-                    file=sys.stderr,
-                )
-                migrate(project)
-            except Exception as e:
-                raise RuntimeError(f"Failed to apply migration {destination}.") from e
-            else:
-                _update_project_config(project, schema_version=destination)
-                print("OK", file=sys.stderr)
-                yield origin, destination
+    from ..project import Project
+
+    if isinstance(root_directory, Project):
+        warnings.warn(
+            "Migrations should be applied to a directory containing a signac project, "
+            "not a project object.",
+            FutureWarning,
+        )
+        root_directory = root_directory.root_directory()
+    lock = FileLock(os.path.join(root_directory, FN_MIGRATION_LOCKFILE))
+    try:
+        with lock:
+            for (origin, destination), migrate in _collect_migrations(root_directory):
+                try:
+                    print(
+                        f"Applying migration for version {origin} to {destination}... ",
+                        end="",
+                        file=sys.stderr,
+                    )
+                    migrate(root_directory)
+                except Exception as e:
+                    raise RuntimeError(
+                        f"Failed to apply migration {destination}."
+                    ) from e
+                else:
+                    config = get_config(os.path.join(root_directory, "signac.rc"))
```

Comment:
Wondering how this would be affected by suggestions in #527 to move to `.signac/config`.

Reply:
Good catch, I missed this change. This line should be using the `_CONFIG_LOADERS` dict just like the other changes in this PR.